src/tools/kata-ctl/src/main.rs

```diff
@@ -35,12 +57,19 @@ fn real_main() -> Result<()> {
         Commands::Iptables(args) => handle_iptables(args),
         Commands::Metrics(args) => handle_metrics(args),
         Commands::Version => handle_version(),
+    };
+
+    // Log errors here, then let the logger go out of scope in main() to ensure
+    // the asynchronous drain flushes all messages before exit()
+    if let Err(ref e) = res {
+        error!(sl!(), "{:#?}", e);
     }
+
+    return res;
 }
 
 fn main() {
-    if let Err(e) = real_main() {
-        eprintln!("ERROR: {:#?}", e);
+    if let Err(_e) = real_main() {
```

Comment:
In `real_main` you print the error, I mean that is not the Rust way of error handling.

Conventionally, you return an error in `real_main`, handle it in `main` and print it out.
Like
```Rust
fn main() {
    if let Err(e) = real_main() {
        error!("{:?}", e);
    }
}
```

Reply:
Agreed - and we tried to get this to work. However - the logger is only in scope in the real_main() function - when it goes out of scope, putting this error! print in main no longer works. I believe there are only three options: 
- The way it is now, with the print in real_main while the logger is still in scope. 
- Revert this specific print to not use slog
- Pull enough of the argument parsing logic into this function so that the logger can be set up here, so the print is valid

Definitely open to other suggestions - this is ugly, and we added a comment about it. Let me know if that comment needs additional detail.

Reply:
I think it's fine to leave it like this. The only way out is to initialize the logger before your call to `real_main()`, but thats not better lol.